File: clients/js/examples/05_aggregation.js
// 05_aggregation.js — top-N customers by spend.
//   $match → $group → $sort → $limit pipeline.
const { OxiDB } = require("../index.js");

(async () => {
  const db = new OxiDB("http://127.0.0.1:9080");
  const orders = db.collection("orders");

  await orders.delete({});
  await orders.insertMany([
    { customer: "alice", amount: 120.50, status: "paid" },
    { customer: "bob",   amount:  45.00, status: "paid" },
    { customer: "alice", amount: 999.00, status: "paid" },
    { customer: "carol", amount: 320.00, status: "paid" },
    { customer: "dan",   amount:  10.00, status: "refunded" },
    { customer: "alice", amount:  50.00, status: "paid" },
    { customer: "bob",   amount: 175.00, status: "paid" },
  ]);

  const top = await orders.aggregate([
    { $match: { status: "paid" } },
    { $group: {
        _id: "$customer",
        spend: { $sum: "$amount" },
        n:     { $sum: 1 },
    }},
    { $sort:  { spend: -1 } },
    { $limit: 3 },
  ]);

  console.log("Top 3 customers:");
  top.forEach((r, i) =>
    console.log(`  ${i + 1}. ${r._id}: $${r.spend} (${r.n} orders)`));
})().catch((e) => { console.error(e); process.exit(1); });

File: clients/julia/examples/embedded/05_aggregation.jl
#!/usr/bin/env julia
# 05_aggregation.jl — group / sum / sort with the aggregation pipeline,
# running entirely in-process.

using OxiDbEmbedded

db = open_db(mktempdir())
insert_many(db, "orders", [
    Dict("category" => "electronics", "amount" => 200, "status" => "completed"),
    Dict("category" => "books",       "amount" => 50,  "status" => "completed"),
    Dict("category" => "electronics", "amount" => 300, "status" => "completed"),
    Dict("category" => "books",       "amount" => 30,  "status" => "pending"),
    Dict("category" => "clothing",    "amount" => 100, "status" => "completed"),
])

println("revenue by category (completed orders only):")
for row in aggregate(db, "orders", [
    Dict("\$match" => Dict("status" => "completed")),
    Dict("\$group" => Dict("_id"   => "\$category",
                           "total" => Dict("\$sum" => "\$amount"),
                           "count" => Dict("\$sum" => 1))),
    Dict("\$sort"  => Dict("total" => -1)),
])
    println("  ", rpad(row["_id"], 12), " \$", row["total"], "  (", row["count"], " orders)")
end

close(db)
